### PR TITLE
impl multiplication for polyval

### LINKED ISSUE #25: feat: GHASH - implement galois multiplication
- impl galois polynomial multiplication subroutines (for ghash and polyval) - note that these will either be the same method, parameterized over the galoi polynomail (maybe too annoying in circom) or two very similar methods, `ghash_mul` and `polyval_mul` with the only difference as the specified irreducible poly's.
    - efficient bitwise galois polynomial mul is well studied, details certainly already exist elsewhere, albeit perhaps not in circom, gather resources on same 
    - impl tests for mult as well
    - impl ghash poly-mul 
    - impl polyval poly-mul

### PR DESCRIPTION
**Pull request by thor314**
Closes #25.

Basing multiplication of the [RustCrytpo](https://github.com/RustCrypto/universal-hashes/blob/master/polyval/src/backend/soft64.rs) software implementation. This PR is about halfway finished, implements `bmul64`, and most of the main galois `mul` subroutine. 

Still need to:
- [x] create a test vector (and hope for the bit-twiddling best)
- [x] impl 64-bit overflowing multiplication 
- [x] finish the last chunk in `mul`

there's a broken test I wrote that isn't actually supposed to pass (and i have no test vectors for it yet; I might generate some from rust-crypto), but triggers circom's compiler errors, which has been useful (if painful).

### CODE REVIEW

File: circuits/aes-gcm/gfmul.circom [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+// > bit-reversed result over 255 bits, not 256. The BearSSL code ends up
+// > with a 256-bit result in zw[], and that value is shifted by one bit,
+// > because of that reversed convention issue. Thus, the code must
+// > include a shifting step to put it back where it should
+//
+// This shift is unnecessary for POLYVAL and has been removed.
+//
+// ref: https://github.com/RustCrypto/universal-hashes/blob/master/polyval/src/backend/soft64.rs#L151
+template MUL() {
+    signal input a[2][64];
```
Comment (devloper) [inline circuits/aes-gcm/gfmul.circom:25]: What's the intuition behind doing [2][64] vs [128]? Given each is represented as a field element internally, my hunch is we don't gain efficiency out of this (tho not certain) while having worse readability.

**Review by devloper — COMMENTED** [review]
Damn!! This is good work! It's a tricky algorithm.  One thought is that his algo appears to be optimized for CPU execution, but I wonder if that maps over to snark efficiency. 

Regardless, starting with a working algorithm seems like the right idea. 

Added only one comment, I'll wait until the PR is out of draft for further review.

File: circuits/aes-gcm/gfmul.circom (hunk 1 above)
Comment (Autoparallel) [reply circuits/aes-gcm/gfmul.circom:25]: maybe to use `REV64()`?

**Comment by thor314** [comment]
All gfmul tests should be logically passing, but I'm having issues getting circomkit to correctly recognize the out[2][64] outputs, seems to only be capturing 64 of the 128 bits, causing these tests to fail.

Only thing left blocking this PR is figuring out that circomkit quirk and we're flying.

@Autoparallel @0xJepsen would you take a look?

**Review by 0xJepsen — APPROVED** [review]
lgtm

**Pull request merged**